### PR TITLE
expression: pushdown substring/char_length to tikv

### LINKED ISSUE #31125: Pushdown substring/substr/mid/char_length to TiKV
## Enhancement
As title said

### PR DESCRIPTION
**Pull request by guo-shaoge**
Signed-off-by: guo-shaoge <shaoge1994@163.com>

### What problem does this PR solve?

Issue Number: close #31125 #30200

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
expression: pushdown substring/substr/mid/char_length to tikv
```

### CODE REVIEW

File: expression/expression.go [hunk 1]
```diff
@@ -966,7 +966,7 @@ func scalarExprSupportedByTiKV(sf *ScalarFunction) bool {
 
 		// string functions.
 		ast.Length, ast.BitLength, ast.Concat, ast.ConcatWS /*ast.Locate,*/, ast.Replace, ast.ASCII, ast.Hex,
-		ast.Reverse, ast.LTrim, ast.RTrim /*ast.Left,*/, ast.Strcmp, ast.Space, ast.Elt, ast.Field,
+		ast.Reverse, ast.LTrim, ast.RTrim /*ast.Left,*/, ast.Strcmp, ast.Space, ast.Elt, ast.Field, ast.Substring,
```
Comment (wjhuang2016) [inline expression/expression.go:969]: How about MID() and SubStr() ?

**Comment by guo-shaoge** [comment]
Related tikv pr: https://github.com/tikv/tikv/pull/11494

File: expression/expression.go (hunk 1 above)
Comment (guo-shaoge) [reply expression/expression.go]: done

File: expression/integration_test.go [hunk 2]
```diff
@@ -3472,36 +3472,36 @@ func TestExprPushdown(t *testing.T) {
 		"(4,'511111','611',7,8,9),(5,'611111','711',8,9,10)")
 
 	// case 1, index scan without double read, some filters can not be pushed to cop task
-	rows := tk.MustQuery("explain format = 'brief' select col2, col1 from t use index(key1) where col2 like '5%' and substr(col1, 1, 1) = '4'").Rows()
+	rows := tk.MustQuery("explain format = 'brief' select col2, col1 from t use index(key1) where col2 like '5%' and from_base64(to_base64(substr(col1, 1, 1))) = '4'").Rows()
 	require.Equal(t, "root", fmt.Sprintf("%v", rows[1][2]))
```
Comment (guo-shaoge) [inline expression/integration_test.go:3476]: `from_base64` and `to_base64` has no plan to pushdown for now...

Comment (ichn-hu) [reply expression/integration_test.go:3476]: this looks so dirty...

**Review by ichn-hu — APPROVED** [review]

**Review by wshwsh12 — APPROVED** [review]

**Pull request merged**